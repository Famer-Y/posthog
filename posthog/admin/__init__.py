from django.contrib import admin

from posthog.admin.admins import (
    AsyncDeletionAdmin,
    CohortAdmin,
    DashboardAdmin,
    DataWarehouseTableAdmin,
    ExperimentAdmin,
    FeatureFlagAdmin,
    InsightAdmin,
    InstanceSettingAdmin,
    OrganizationAdmin,
    PersonAdmin,
    PersonDistinctIdAdmin,
    PluginAdmin,
    PluginConfigAdmin,
    ProjectAdmin,
    RedisAdmin,
    SurveyAdmin,
    TeamAdmin,
    TextAdmin,
    UserAdmin,
)
from posthog.models import (
    AsyncDeletion,
    Cohort,
    Dashboard,
    DataWarehouseTable,
    Experiment,
    FeatureFlag,
    Insight,
    InstanceSetting,
    Organization,
    Person,
    PersonDistinctId,
    Plugin,
    PluginConfig,
    Project,
    RedisMutation,
    Survey,
    Team,
    Text,
    User,
)

admin.site.register(Organization, OrganizationAdmin)
admin.site.register(Project, ProjectAdmin)
admin.site.register(Team, TeamAdmin)
admin.site.register(User, UserAdmin)

admin.site.register(Dashboard, DashboardAdmin)
admin.site.register(Insight, InsightAdmin)

admin.site.register(Experiment, ExperimentAdmin)
admin.site.register(FeatureFlag, FeatureFlagAdmin)

admin.site.register(AsyncDeletion, AsyncDeletionAdmin)
admin.site.register(InstanceSetting, InstanceSettingAdmin)
admin.site.register(PluginConfig, PluginConfigAdmin)
admin.site.register(Plugin, PluginAdmin)
admin.site.register(Text, TextAdmin)

admin.site.register(Cohort, CohortAdmin)
admin.site.register(Person, PersonAdmin)
admin.site.register(PersonDistinctId, PersonDistinctIdAdmin)

admin.site.register(Survey, SurveyAdmin)

admin.site.register(DataWarehouseTable, DataWarehouseTableAdmin)

admin.site.register(RedisMutation, RedisAdmin)
