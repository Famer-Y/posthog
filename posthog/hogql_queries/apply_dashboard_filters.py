from sentry_sdk import capture_exception
from posthog.hogql_queries.query_runner import get_query_runner
from posthog.models import Team
from posthog.schema import DashboardFilter, NodeKind

DATA_TABLE_LIKE_NODE_KINDS = [NodeKind.DataTableNode, NodeKind.DataVisualizationNode]


# Apply the filters from the django-style Dashboard object
def apply_dashboard_filters_to_dict(query: dict, filters: dict, team: Team) -> dict:
    if not filters:
        return query

    if query.get("kind") in DATA_TABLE_LIKE_NODE_KINDS:
        source = apply_dashboard_filters_to_dict(query["source"], filters, team)
        return {**query, "source": source}

    try:
        query_runner = get_query_runner(query, team)
    except ValueError:
        capture_exception()
        return query
    return query_runner.apply_dashboard_filters(DashboardFilter(**filters)).dict()
