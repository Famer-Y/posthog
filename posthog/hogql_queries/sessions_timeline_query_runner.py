from collections import defaultdict
from datetime import timedelta
import json
from typing import DefaultDict, Dict, List, Optional, Any, cast
from posthog.api.element import ElementSerializer


from posthog.clickhouse.client.connection import Workload
from posthog.hogql import ast
from posthog.hogql.parser import parse_select
from posthog.hogql.query import execute_hogql_query
from posthog.hogql.timings import HogQLTimings
from posthog.hogql_queries.query_runner import QueryRunner
from posthog.models import Team
from posthog.models.element.element import chain_to_elements
from posthog.schema import EventType, SessionsTimelineQuery, SessionsTimelineQueryResponse, TimelineEntry


class SessionsTimelineQueryRunner(QueryRunner):
    query: SessionsTimelineQuery
    query_type = SessionsTimelineQuery

    def __init__(
        self,
        query: SessionsTimelineQuery | Dict[str, Any],
        team: Team,
        timings: Optional[HogQLTimings] = None,
    ):
        super().__init__(query, team, timings)
        if isinstance(query, SessionsTimelineQuery):
            self.query = query
        else:
            self.query = SessionsTimelineQuery.model_validate(query)

    def to_query(self) -> ast.SelectQuery:
        if self.timings is None:
            self.timings = HogQLTimings()

        with self.timings.measure("build_ast"):
            select_query = cast(
                ast.SelectQuery,
                parse_select(
                    """
                    WITH (
                        SELECT DISTINCT $session_id
                        FROM events
                        WHERE events.timestamp > toDateTime({after}) AND events.timestamp <= toDateTime({before})
                        LIMIT 100
                    ) AS relevant_session_ids
                    SELECT uuid, $session_id, timestamp, event, properties, distinct_id, elements_chain
                    FROM events
                    WHERE
                        events.timestamp >= (
                            SELECT min(timestamp)
                            FROM events
                            WHERE events.$session_id IN relevant_session_ids OR (
                                $session_id IS NULL
                                AND events.timestamp > toDateTime({after}) AND events.timestamp < toDateTime({before})
                            )
                        )
                        AND events.timestamp <= (
                            SELECT max(timestamp)
                            FROM events
                            WHERE events.$session_id IN relevant_session_ids OR (
                                $session_id IS NULL
                                AND events.timestamp > toDateTime({after}) AND events.timestamp < toDateTime({before})
                            )
                        )
                    ORDER BY timestamp ASC""",
                    placeholders={
                        "before": ast.Constant(value=self.query.before),
                        "after": ast.Constant(value=self.query.after),
                    },
                ),
            )
            assert select_query.ctes is not None
            assert isinstance(select_query.ctes["relevant_session_ids"].expr, ast.SelectQuery)
            if self.query.personId:
                select_query.ctes["relevant_session_ids"].expr.where = ast.CompareOperation(
                    left=ast.Field(chain=["person_id"]),
                    right=ast.Constant(value=self.query.personId),
                    op=ast.CompareOperationOp.Eq,
                )
                select_query.where = ast.CompareOperation(
                    left=ast.Field(chain=["person_id"]),
                    right=ast.Constant(value=self.query.personId),
                    op=ast.CompareOperationOp.Eq,
                )
            return select_query

    def calculate(self) -> SessionsTimelineQueryResponse:
        query_result = execute_hogql_query(
            query=self.to_query(),
            team=self.team,
            workload=Workload.ONLINE,
            query_type="SessionsTimelineQuery",
            timings=self.timings,
        )
        assert query_result.results is not None
        timeline_entries_map: DefaultDict[str, List[EventType]] = defaultdict(list)
        for (
            uuid,
            session_id,
            timestamp_parsed,
            event,
            properties_raw,
            distinct_id,
            elements_chain,
        ) in query_result.results:
            timeline_entries_map[session_id].append(
                EventType(
                    id=str(uuid),
                    distinct_id=distinct_id,
                    event=event,
                    timestamp=timestamp_parsed.isoformat(),
                    properties=json.loads(properties_raw),
                    elements=ElementSerializer(chain_to_elements(elements_chain), many=True).data,
                )
            )
        for events in timeline_entries_map.values():
            events.reverse()
        timeline_entries = [
            TimelineEntry(sessionId=session_id, events=events)
            for session_id, events in reversed(timeline_entries_map.items())
        ]

        return SessionsTimelineQueryResponse(
            results=timeline_entries,
            hasMore=False,  # TODO
            timings=self.timings.to_list(),
            hogql=query_result.hogql,
        )

    def _is_stale(self, cached_result_package):
        return True

    def _refresh_frequency(self):
        return timedelta(minutes=1)
