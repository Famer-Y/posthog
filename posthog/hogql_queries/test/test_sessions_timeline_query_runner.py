from posthog.hogql_queries.sessions_timeline_query_runner import SessionsTimelineQueryRunner
from posthog.schema import EventType, SessionsTimelineQuery, TimelineEntry
from posthog.test.base import APIBaseTest, ClickhouseTestMixin, snapshot_clickhouse_queries
from posthog.test.test_journeys import journeys_for


class TestSessionsTimelineQueryRunner(ClickhouseTestMixin, APIBaseTest):
    def _create_runner(self, query: SessionsTimelineQuery) -> SessionsTimelineQueryRunner:
        return SessionsTimelineQueryRunner(team=self.team, query=query)

    @snapshot_clickhouse_queries
    def test_simple_sessions_unfiltered(self):
        journeys_for(
            team=self.team,
            events_by_person={
                "person1": [
                    # The sessions are sorted most recently started to least,
                    # while events within most recent to least recent
                    {
                        "event_uuid": "6e6e645b-2936-4613-b409-b33f4d9a0f18",
                        "event": "$pageview",
                        "timestamp": "2021-01-01 12:00:05",
                        "properties": {"$session_id": "s1"},
                    },
                    {
                        "event_uuid": "c15119f2-b243-4547-ab46-1b29a0435948",
                        "event": "user signed up",
                        "timestamp": "2021-01-01 13:00:05",
                        "properties": {"$session_id": "s1"},
                    },
                    {
                        "event_uuid": "b826a13e-aae3-4766-b407-0d3a582140e4",
                        "event": "$pageview",
                        "timestamp": "2021-01-01 14:00:05",
                        "properties": {"$session_id": "s1"},
                    },
                    {
                        "event_uuid": "e1208e6b-8101-4dde-ba21-c47781bb5bad",
                        "event": "$pageview",
                        "timestamp": "2021-01-01 17:00:05",
                        "properties": {"$session_id": "s2"},
                    },
                ],
                "person2": [  # Partly overlapping with person1
                    {
                        "event_uuid": "605f6843-bf83-4d7b-b9a0-4d6f7f57415f",
                        "event": "$pageview",
                        "timestamp": "2021-01-01 13:30:00",
                        "properties": {"$session_id": "s3"},
                    },
                    {
                        "event_uuid": "04dde300-a6c3-4372-9366-80472c2d02b1",
                        "event": "did important thing",
                        "timestamp": "2021-01-02 02:00:00",
                        "properties": {"$session_id": "s3"},
                    },
                ],
            },
        )

        runner = self._create_runner(SessionsTimelineQuery(before="2021-01-01T18:00:00Z", after="2021-01-01T06:00:00Z"))
        response = runner.calculate()

        assert response.results == [
            TimelineEntry(
                sessionId="s2",
                events=[
                    EventType(
                        id="e1208e6b-8101-4dde-ba21-c47781bb5bad",
                        distinct_id="person1",
                        event="$pageview",
                        timestamp="2021-01-01T17:00:05+00:00",
                        properties={"$session_id": "s2"},
                    )
                ],
            ),
            TimelineEntry(
                sessionId="s3",
                events=[
                    EventType(
                        id="04dde300-a6c3-4372-9366-80472c2d02b1",
                        distinct_id="person2",
                        event="did important thing",
                        timestamp="2021-01-02T02:00:00+00:00",
                        properties={"$session_id": "s3"},
                    ),
                    EventType(
                        id="605f6843-bf83-4d7b-b9a0-4d6f7f57415f",
                        distinct_id="person2",
                        event="$pageview",
                        timestamp="2021-01-01T13:30:00+00:00",
                        properties={"$session_id": "s3"},
                    ),
                ],
            ),
            TimelineEntry(
                sessionId="s1",
                events=[
                    EventType(
                        id="b826a13e-aae3-4766-b407-0d3a582140e4",
                        distinct_id="person1",
                        event="$pageview",
                        timestamp="2021-01-01T14:00:05+00:00",
                        properties={"$session_id": "s1"},
                    ),
                    EventType(
                        id="c15119f2-b243-4547-ab46-1b29a0435948",
                        distinct_id="person1",
                        event="user signed up",
                        timestamp="2021-01-01T13:00:05+00:00",
                        properties={"$session_id": "s1"},
                    ),
                    EventType(
                        id="6e6e645b-2936-4613-b409-b33f4d9a0f18",
                        distinct_id="person1",
                        event="$pageview",
                        timestamp="2021-01-01T12:00:05+00:00",
                        properties={"$session_id": "s1"},
                    ),
                ],
            ),
        ]

    # TODO: Test with personId
    # TODO: Test with out of session events
    # TODO: Test with overlapping sessions
