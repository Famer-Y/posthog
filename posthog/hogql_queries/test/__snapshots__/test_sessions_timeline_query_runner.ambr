# name: TestSessionsTimelineQueryRunner.test_simple_sessions_unfiltered
  '
  SELECT events.uuid,
         events.`$session_id`,
         toTimeZone(events.timestamp, 'UTC'),
         events.event,
         events.properties,
         events.distinct_id
  FROM events
  WHERE and(equals(events.team_id, 2), greaterOrEquals(toTimeZone(events.timestamp, 'UTC'),
                                                         (SELECT min(toTimeZone(events.timestamp, 'UTC'))
                                                          FROM events
                                                          WHERE and(equals(events.team_id, 2), or(in(events.`$session_id`,
                                                                                                       (SELECT DISTINCT events.`$session_id`
                                                                                                        FROM events
                                                                                                        WHERE and(equals(events.team_id, 2), greater(toTimeZone(events.timestamp, 'UTC'), parseDateTime64BestEffortOrNull('2021-01-01T06:00:00Z', 6, 'UTC')), lessOrEquals(toTimeZone(events.timestamp, 'UTC'), parseDateTime64BestEffortOrNull('2021-01-01T18:00:00Z', 6, 'UTC')))
                                                                                                        LIMIT 100)), and(isNull(events.`$session_id`), greater(toTimeZone(events.timestamp, 'UTC'), parseDateTime64BestEffortOrNull('2021-01-01T06:00:00Z', 6, 'UTC')), less(toTimeZone(events.timestamp, 'UTC'), parseDateTime64BestEffortOrNull('2021-01-01T18:00:00Z', 6, 'UTC'))))))), lessOrEquals(toTimeZone(events.timestamp, 'UTC'),
                                                                                                                                                                                                                                                                                                                                                                                                          (SELECT max(toTimeZone(events.timestamp, 'UTC'))
                                                                                                                                                                                                                                                                                                                                                                                                           FROM events
                                                                                                                                                                                                                                                                                                                                                                                                           WHERE and(equals(events.team_id, 2), or(in(events.`$session_id`,
                                                                                                                                                                                                                                                                                                                                                                                                                                                        (SELECT DISTINCT events.`$session_id`
                                                                                                                                                                                                                                                                                                                                                                                                                                                         FROM events
                                                                                                                                                                                                                                                                                                                                                                                                                                                         WHERE and(equals(events.team_id, 2), greater(toTimeZone(events.timestamp, 'UTC'), parseDateTime64BestEffortOrNull('2021-01-01T06:00:00Z', 6, 'UTC')), lessOrEquals(toTimeZone(events.timestamp, 'UTC'), parseDateTime64BestEffortOrNull('2021-01-01T18:00:00Z', 6, 'UTC')))
                                                                                                                                                                                                                                                                                                                                                                                                                                                         LIMIT 100)), and(isNull(events.`$session_id`), greater(toTimeZone(events.timestamp, 'UTC'), parseDateTime64BestEffortOrNull('2021-01-01T06:00:00Z', 6, 'UTC')), less(toTimeZone(events.timestamp, 'UTC'), parseDateTime64BestEffortOrNull('2021-01-01T18:00:00Z', 6, 'UTC'))))))))
  ORDER BY toTimeZone(events.timestamp, 'UTC') ASC
  LIMIT 100 SETTINGS readonly=2,
                     max_execution_time=60,
                     allow_experimental_object_type=1
  '
---
# name: TestSessionsTimelineQueryRunner.test_simple_sessions_with_person_id
  '
  SELECT events.uuid,
         events.`$session_id`,
         toTimeZone(events.timestamp, 'UTC'),
         events.event,
         events.properties,
         events.distinct_id
  FROM events
  INNER JOIN
    (SELECT argMax(person_distinct_id2.person_id, person_distinct_id2.version) AS person_id,
            person_distinct_id2.distinct_id AS distinct_id
     FROM person_distinct_id2
     WHERE equals(person_distinct_id2.team_id, 2)
     GROUP BY person_distinct_id2.distinct_id
     HAVING ifNull(equals(argMax(person_distinct_id2.is_deleted, person_distinct_id2.version), 0), 0)) AS events__pdi ON equals(events.distinct_id, events__pdi.distinct_id)
  WHERE and(equals(events.team_id, 2), ifNull(equals(events__pdi.person_id, '018b4922-f81a-0000-7a5a-2412961cdcd1'), 0))
  ORDER BY toTimeZone(events.timestamp, 'UTC') ASC
  LIMIT 100 SETTINGS readonly=2,
                     max_execution_time=60,
                     allow_experimental_object_type=1
  '
---
